### DISCUSSION TITLE
Provide panel relative time macros

### DISCUSSION THREAD

**Discussion by imReker** [grafana/grafana-infinity-datasource#167] [Q&A] [answered]
Global time variable `$__from` and `$__to` are owned by dashboard, not panel, so currently there is no way to get panel's real time when panel has set 'Relative time' or 'Time shift' settings.

**Comment by yesoreyeram** [comment]
Panel have access to those variables. Wasn't working for you?

**Reply by imReker** [reply]
> Panel have access to those variables. Wasn't working for you?

No, `$__from` is owned by dashboard, so it's same to all panels in the dashboard.
But panel can have it's own relative time, which may totally different from time of dashboard.
![1](https://user-images.githubusercontent.com/4949937/136653309-63793827-ac8e-4346-a07b-213b67d2f1f1.png)

**Reply by yesoreyeram** [reply]
You are referring to variable in panel title which is outside plugin's scope. So this must be fixed by grafana.

**Reply by imReker** [reply]
> You are referring to variable in panel title which is outside plugin's scope. So this must be fixed by grafana.

Nope, other plugins such as prometheus can get this relative time

**Reply by yesoreyeram** [reply]
In such case, can you confirm that by changing the datasource in above screenshot please? 

AFAIK, datasource itself receive timerange from dashboard. In other words, datasource doesn't control timerange

**Reply by tris** [reply]
I'm having the same issue.  I have a dashboard where _most_ of the panels follow the "last 3 hours" dashboard-level time range, but I have one panel that needs to be looking 2 days ahead, and timeshifted.  Using `${__from:date:YYYYMMDD}` and `${__to:date:YYYYMMDD}` gives me ~today's date for both values, so the endpoint I'm using doesn't return the data I need from the future.

To be very explicit about my use case, this is for displaying a tide chart:
`https://api.tidesandcurrents.noaa.gov/api/prod/datagetter?product=predictions&time_zone=lst_ldt&units=english&interval=hilo&format=json&datum=MLLW&station=9414290&begin_date=${__from:date:YYYYMMDD}&end_date=${__to:date:YYYYMMDD}`

This _seems_ to be a Grafana issue (see grafana/grafana#20147 and grafana/grafana#38954) and would ideally be solved there.  (@imReker actually opened that second one.)

But, there's some commentary suggesting that this already works properly in the MySQL datasource (using its `$__timeFrom()`, `$__timeTo()`, `$__timeFilter()`, etc.), so perhaps it can be fixed similarly in Infinity?

**Reply by PlatypussDivva** [reply]
I wanted to upvote this request.  

@yesoreyeram Yes the panel with the infinity datasource has access to the global $__to and $__from variables and works as expected when using those.

The overall issue is that panels have Query Options -> relative time and Query Options -> time shift, intended to overwrite the globally defined time range.  

You would expect that this is provided by Grafana core and that when you use the query options in a panel - any reference to $__to and $__from inside that panel would use the override but that is not the case.  $__to and $_from will always have the global dashboard-level value.  And it is up to datasource plugins to provide a macro that uses the panels query option overrides. 

Is there a way to get something like $panelTo and $panelFrom implemented?     For example MySQL datasource (using its $__timeFrom(), $__timeTo(), $__timeFilter(), etc.) which can be used in the queries and will honor the relative time and time shift data.

**Reply by vertebarbe** [reply]
If anyone found a workaround for this yet, please share it here.

![image](https://github.com/grafana/grafana-infinity-datasource/assets/35434346/fa6708b7-0b7e-441f-b7ec-22603380eea8)
I can see that the "from" and "to" fields inside the request object in the query inspector of the infinity datasource show the correct panel-relative timestamps, maybe this could easily be returned by a macro for use in query parameters, or maybe there is another way to get these values?

**Comment by misilot** [comment]
Did anyone find a solution for this?

I am trying to allow the user select the date range they want and pull that data but also pull the prior years data so they can compare it.

So if a user selects

2024-01-01 00:00:00 to 2024-03-01 00:00:00 the relative from a year ago should be
2023-01-01 00:00:00 to 2023-03-01 00:00:00

**Reply by misilot** [reply]
I did create this [issue](https://github.com/grafana/grafana/issues/83803) in Grafana, but not sure if it could be fixed in the datasource first?

**Comment by yesoreyeram** [comment]
Sorry it slipped through. As @tris 
and  @PlatypussDivva pointed out, this is something needs to be fixed at grafana level. Data source plugin can't do anything here. But similar to MySQL, the infinity can provide additional time macro to overcome this issue. Workaround `timeFrom` and `timeTo` macros coming in next few version.

**Reply by yesoreyeram** [reply]
created issue for tracking https://github.com/grafana/grafana-infinity-datasource/issues/796

**Comment by yesoreyeram** [comment, marked as answer]
you can now use time macros as specified in https://grafana.com/docs/plugins/yesoreyeram-infinity-datasource/latest/query/macros/#time-macros-__timefrom-and-__timeto which will be interpolated in the backend and respect panel relative time.

**Reply by yosiasz** [reply]
I am not seeing any examples that show use of these in date ranges filtering either for csv back end parser or uql
